market/ipni/ipni-provider/spark.go

```diff
@@ -0,0 +1,230 @@
+package ipni_provider
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	eabi "github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	fbig "github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/builtin"
+
+	"github.com/filecoin-project/curio/market/ipni/spark"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/build/buildconstants"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+func (p *Provider) updateSparkContract(ctx context.Context) error {
+	var priv []byte
+	var onChainPeerID string
+
+	err := p.db.QueryRow(ctx, `select priv_key, peer_id from libp2p LIMIT 1`).Scan(&priv, &onChainPeerID)
+	if err != nil {
+		return xerrors.Errorf("querying libp2p peer: %w", err)
+	}
+
+	pKey, err := crypto.UnmarshalPrivateKey(priv)
+	if err != nil {
+		return xerrors.Errorf("unmarshaling private key: %w", err)
+	}
+
+	for _, pInfo := range p.keys {
+		pInfo := pInfo
+		mInfo, err := p.full.StateMinerInfo(ctx, pInfo.Miner, types.EmptyTSK)
+		if err != nil {
+			return err
+		}
+
+		if mInfo.PeerId == nil {
+			return xerrors.Errorf("peer id not found for miner: %s", pInfo.Miner)
+		}
+
+		if mInfo.PeerId.String() != onChainPeerID {
+			return xerrors.Errorf("peer id mismatch for miner: %s: onChain: %s and DB: %s", pInfo.Miner, mInfo.PeerId.String(), onChainPeerID)
+		}
+
+		contractID, err := spark.GetContractAddress()
+		if err != nil {
+			return err
+		}
+
+		to, err := ethtypes.ParseEthAddress(contractID)
+		if err != nil {
+			return xerrors.Errorf("failed to parse contract address: %w", err)
+		}
+
+		toAddr, err := to.ToFilecoinAddress()
+		if err != nil {
+			return xerrors.Errorf("failed to convert Eth address to Filecoin address: %w", err)
+		}
+
+		// Parse the contract ABI
+		parsedABI, err := eabi.JSON(strings.NewReader(spark.GetPeerAbi))
+		if err != nil {
+			log.Fatalf("Failed to parse getPeer ABI: %v", err)
+		}
+
+		// Encode the function call
+		callData, err := parsedABI.Pack("getPeerData", pInfo.SPID)
+		if err != nil {
+			log.Fatalf("Failed to pack function call data: %v", err)
```

Comment:
Return error, don't do fatal which will crash the node

Reply:
Fixed